prow/cluster_lib.sh

```diff
@@ -20,23 +20,147 @@
 #             e2e-suite               #
 #                                     #
 #######################################
-KUBE_USER="istio-prow-test-job@istio-testing.iam.gserviceaccount.com"
+
+KUBE_USER="${KUBE_USER:-istio-prow-test-job@istio-testing.iam.gserviceaccount.com}"
+SETUP_CLUSTERREG=${SETUP_CLUSTERREG:-False}
+USE_GKE=${USE_GKE:-True}
+CLUSTER_NAME=
+SA_NAMESPACE=istio-system-multi
+
+
+function gen_kubeconf_from_sa () {
+    local service_account=$1
+    local filename=$2
+
+    NAMESPACE=${SA_NAMESPACE:-istio-system}
+
+    SERVER=$(kubectl config view --minify=true -o "jsonpath={.clusters[].cluster.server}")
+    SECRET_NAME=$(kubectl get sa ${service_account} -n ${NAMESPACE} -o jsonpath='{.secrets[].name}')
+    CA_DATA=$(kubectl get secret ${SECRET_NAME} -n ${NAMESPACE} -o "jsonpath={.data['ca\.crt']}")
+    TOKEN=$(kubectl get secret ${SECRET_NAME} -n ${NAMESPACE} -o "jsonpath={.data['token']}" | base64 --decode)
+
+    cat <<EOF > ${filename}
+      apiVersion: v1
+      clusters:
+         - cluster:
+             certificate-authority-data: ${CA_DATA}
+             server: ${SERVER}
+           name: ${CLUSTER_NAME}
+      contexts:
+         - context:
+             cluster: ${CLUSTER_NAME}
+             user: ${CLUSTER_NAME}
+           name: ${CLUSTER_NAME}
+      current-context: ${CLUSTER_NAME}
+      kind: Config
+      preferences: {}
+      users:
+         - name: ${CLUSTER_NAME}
+           user:
+             token: ${TOKEN}
+EOF
+
+}
+
+function setup_clusterreg () {
+    # setup cluster-registries dir setup by mason
+    CLUSTERREG_DIR=${CLUSTERREG_DIR:-$(mktemp -d /tmp/clusterregXXX)}
+
+    SERVICE_ACCOUNT=istio-multi
+
+    # mason dumps all the kubeconfigs into the same file but we need to use per cluster
+    # files for the clusterregsitry config.  Create the separate files.
+    #  -- if PILOT_CLUSTER not set, assume pilot install to be in the first cluster
+    PILOT_CLUSTER=${PILOT_CLUSTER:-$(kubectl config current-context)}
+    unset IFS
+    k_contexts=$(kubectl config get-contexts -o name)
+    for context in $k_contexts; do
+        if [[ "$PILOT_CLUSTER" != "$context" ]]; then
+            kubectl config use-context ${context}
+             
+            kubectl create ns ${SA_NAMESPACE}
+            kubectl create sa ${SERVICE_ACCOUNT} -n ${SA_NAMESPACE}
+            kubectl create clusterrolebinding istio-multi --clusterrole=cluster-admin --serviceaccount=${SA_NAMESPACE}:${SERVICE_ACCOUNT}
+            CLUSTER_NAME=$(kubectl config view --minify=true -o "jsonpath={.clusters[].name}")
+            if [[ "$CLUSTER_NAME" =~ .*"_".* ]]; then
+                # if clustername has '_' set value to stuff after the last '_' due to k8s secret data name limitation
+                CLUSTER_NAME=${CLUSTER_NAME##*_}
+            fi
+            KUBECFG_FILE=${CLUSTERREG_DIR}/${CLUSTER_NAME}
+            gen_kubeconf_from_sa $SERVICE_ACCOUNT $KUBECFG_FILE
```

Comment:
User {} around variable names for consistency.

Reply:
sure